pkg/server/endpoints/entryfetcher_test.go

```diff
@@ -0,0 +1,199 @@
+	assert.NoError(t, err)
+	assert.NotNil(t, ef)
+}
+
+func TestNewAuthorizedEntryFetcherWithFullCacheErrorBuildingCache(t *testing.T) {
+	ctx := context.Background()
+	log := logrus.New()
+	log.Out = ioutil.Discard
+	metrics := fakemetrics.New()
+	ds := fakedatastore.New(t)
+	ds.SetNextError(errors.New("some datastore error"))
+	clk := clock.NewMock()
+
+	ef, err := NewAuthorizedEntryFetcherWithFullCache(ctx, log, metrics, ds, clk)
+	assert.Error(t, err)
+	assert.Nil(t, ef)
+}
+
+func TestFetchRegistrationEntries(t *testing.T) {
+	ctx := context.Background()
+	log := logrus.New()
+	log.Out = ioutil.Discard
+	metrics := fakemetrics.New()
+	ds := fakedatastore.New(t)
+	clk := clock.NewMock()
+	const rootID = "spiffe://example.org/root"
+	expected := setupExpectedEntriesData(ctx, t, ds, rootID)
+
+	ef, err := NewAuthorizedEntryFetcherWithFullCache(ctx, log, metrics, ds, clk)
+	require.NoError(t, err)
+	require.NotNil(t, ef)
+
+	agentSpiffeID, err := spiffeid.FromString(rootID)
+	require.NoError(t, err)
+
+	entries, err := ef.FetchAuthorizedEntries(ctx, agentSpiffeID)
+	assert.NoError(t, err)
+	assert.Equal(t, expected, entries)
+}
+
+func TestRunRebuildCacheTask(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	log := logrus.New()
+	log.Out = ioutil.Discard
+	metrics := fakemetrics.New()
+	ds := fakedatastore.New(t)
+	clk := clock.NewMock()
+	const rootID = "spiffe://example.org/root"
+	agentSpiffeID, err := spiffeid.FromString(rootID)
+	require.NoError(t, err)
+
+	ef, err := NewAuthorizedEntryFetcherWithFullCache(ctx, log, metrics, ds, clk)
+	require.NoError(t, err)
+	require.NotNil(t, ef)
+
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- ef.RunRebuildCacheTask(ctx)
+	}()
+
+	entries, err := ef.FetchAuthorizedEntries(ctx, agentSpiffeID)
+	assert.NoError(t, err)
+	assert.Empty(t, entries)
+
+	expected := setupExpectedEntriesData(ctx, t, ds, rootID)
+
+	// Entries should still not be in the cache yet
+	// because the clock has not elapsed long enough for the cache rebuild task to execute.
+	entries, err = ef.FetchAuthorizedEntries(ctx, agentSpiffeID)
+	assert.NoError(t, err)
+	assert.Empty(t, entries)
+
+	// Verify that rebuild task gracefully handles downstream errors and retries after the reload interval elapses again
+	ds.SetNextError(errors.New("some datastore error"))
+	clk.Add(cacheReloadInterval)
+	entries, err = ef.FetchAuthorizedEntries(ctx, agentSpiffeID)
+	assert.NoError(t, err)
+	assert.Empty(t, entries)
+
+	// When the rebuild task is able to complete successfully,
+	// the cache should now contain the Agent's new authorized entries
+	clk.Add(cacheReloadInterval)
+	entries, err = ef.FetchAuthorizedEntries(ctx, agentSpiffeID)
+	assert.NoError(t, err)
+	assert.Equal(t, expected, entries)
+
+	cancel()
+	assert.NoError(t, <-errChan)
+}
+
+func setupExpectedEntriesData(ctx context.Context, t *testing.T, ds datastore.DataStore, agentSpiffeID string) []*types.Entry {
```

Comment:
like mentioned above, I think a lot of this code can be simplified if you leverage the spiffeid.ID and spiffeid.TrustDomain types